src/WebJobs.Script/IO/AutoRecoveringFileSystemWatcher.cs

```diff
@@ -0,0 +1,202 @@
+
+        public AutoRecoveringFileSystemWatcher(string path, string filter = "*.*", 
+            bool includeSubdirectories = true, WatcherChangeTypes changeTypes = WatcherChangeTypes.All, TraceWriter traceWriter = null)
+        {
+            _path = path;
+            _filter = filter;
+            _changeTypes = changeTypes;
+            _includeSubdirectories = includeSubdirectories;
+            _traceWriter = traceWriter;
+            _cancellationTokenSource = new CancellationTokenSource();
+            _cancellationToken = _cancellationTokenSource.Token;
+            _handleFileError = new Action<ErrorEventArgs>(OnFileWatcherError).Debounce();
+            
+            InitializeWatcher();
+        }
+      
+        ~AutoRecoveringFileSystemWatcher()
+        {
+            Dispose(false);
+        }
+
+        private void InitializeWatcher()
+        {
+            if (!Directory.Exists(_path))
+            {
+                throw new DirectoryNotFoundException($"The path '{_path}' cannot be found.");
+            }
+
+            lock (_syncRoot)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
+
+                _fileWatcher = new FileSystemWatcher(_path, _filter)
+                {
+                    IncludeSubdirectories = _includeSubdirectories,
+                    EnableRaisingEvents = true
+                };
+
+                _fileWatcher.Error += (s, a) => _handleFileError(a);
+                AddEventSubscriptions(_fileWatcher, _changeTypes);
+            }
+        }
+
+        private void AddEventSubscriptions(FileSystemWatcher fileWatcher, WatcherChangeTypes changeTypes)
+        {
+            if (changeTypes.HasFlag(WatcherChangeTypes.Changed))
+            {
+                fileWatcher.Changed += OnFileChanged;
+            }
+
+            if (changeTypes.HasFlag(WatcherChangeTypes.Created))
+            {
+                fileWatcher.Created += OnFileChanged;
+            }
+
+            if (changeTypes.HasFlag(WatcherChangeTypes.Deleted))
+            {
+                fileWatcher.Deleted += OnFileChanged;
+            }
+
+            if (changeTypes.HasFlag(WatcherChangeTypes.Renamed))
+            {
+                fileWatcher.Renamed += OnFileChanged;
+            }
+        }
+
+        private void RemoveEventSubscriptions(FileSystemWatcher fileWatcher)
+        {
+            fileWatcher.Changed -= OnFileChanged;
+            fileWatcher.Created -= OnFileChanged;
+            fileWatcher.Deleted -= OnFileChanged;
+            fileWatcher.Renamed -= OnFileChanged;
+        }
+
+        protected void OnFileChanged(object sender, FileSystemEventArgs e)
+        {
+            Changed?.Invoke(this, e);
+        }
+
+        protected void OnFileWatcherError(ErrorEventArgs obj)
+        {
+            Trace($"Failure detected '{obj.GetExceptionMessage()}'. Initiating recovery...", TraceLevel.Warning);
+
+            Recover().ContinueWith(t =>
```

Comment:
Good point! Wouldn't happen on the recovery path but if we happen to get additional error events (or outside of our debounce window) while recovering that would indeed happen.